### DISCUSSION TITLE
How to priorities dependent requests execution order ?

### DISCUSSION THREAD

**Discussion by rabin999** [pact-foundation/pact-js#1174] [Q&A]
we have 2 microservices.

**About the Issue**
- multiple pact tests which are dependent on each other.
- verifier need to execute those requests based on their priority

**Example**
1. request A -> Create order with quantity 1 with SKU: `abc`
2. request C -> retrieve list of remaining qty for SKU: `abc`
3. request B -> Create another order with quantity 1 with SKU: `abc`

Execution Order of those requests
- 1 -> 3 -> 2

**Comment by portyu9** [comment]
Hi @rabin999 

You should **not depend on Pact executing interactions in a particular order**.

Pact interactions are intentionally independent. The verifier must be able to verify each interaction without relying on a previous interaction having modified the provider.

So instead of requiring:

```text
A: create first order
B: create second order
C: retrieve remaining quantity

required order: A -> B -> C
```

each interaction should define the provider state it needs.

Conceptually:

```text
A
Given SKU abc has quantity 2
When an order for quantity 1 is created
Then ...

B
Given SKU abc has remaining quantity 1
When another order for quantity 1 is created
Then ...

C
Given SKU abc has the expected remaining quantity
When remaining inventory is requested
Then ...
```

In Pact JS you implement those preconditions with provider states/state handlers:

```js
const opts = {
    providerBaseUrl: 'http://localhost:8080',

    stateHandlers: {
        'SKU abc has quantity 2': async () => {
            await inventoryRepository.setQuantity('abc', 2)
        },

        'SKU abc has remaining quantity 1': async () => {
            await inventoryRepository.setQuantity('abc', 1)
        },

        'SKU abc is out of stock': async () => {
            await inventoryRepository.setQuantity('abc', 0)
        }
    }
}

await new Verifier(opts).verifyProvider()
```

The consumer interaction would declare the corresponding state:

```js
provider
    .given('SKU abc has remaining quantity 1')
    .uponReceiving('an order for SKU abc')
    // ...
```

This makes every interaction deterministic regardless of which order the verifier chooses.

If what you actually need to verify is the complete workflow:

```text
create order
→ inventory changes
→ create another order
→ retrieve inventory
→ verify cumulative side effects
```

that is an **integration/functional workflow test**, not a Pact contract test. Keep that scenario in the provider's integration/E2E test suite.

Pact should verify that both sides agree on each individual request/response contract, not orchestrate stateful business workflows by forcing interaction execution order.

Please mark this answer as accepted if it helped, thank you.
